File: TermiRustMobile/Models/ControllerModels.swift
import Foundation

struct HostRoute: Codable, Hashable, Sendable {
    let address: String
    let port: UInt16

    init(address: String, port: UInt16) throws {
        let trimmed = address.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, trimmed.utf8.count <= 255, port > 0 else {
            throw ControllerModelError.invalidRoute
        }
        self.address = trimmed
        self.port = port
    }
}

struct PairedHostRecord: Codable, Identifiable, Hashable, Sendable {
    static let currentSchemaVersion = 1

    let schemaVersion: Int
    let id: String
    let displayName: String
    let route: HostRoute
    let hostStaticPublicKey: Data
    let deviceStaticKeyId: String
    let deviceId: UUID
    let identityGeneration: UInt64
    let revocationEpoch: UInt64
    let sessionGeneration: UInt64
    let capabilityBits: UInt16
    let pairedAt: Date

    init(
        id: String,
        displayName: String,
        route: HostRoute,
        hostStaticPublicKey: Data,
        deviceStaticKeyId: String,
        deviceId: UUID,
        identityGeneration: UInt64,
        revocationEpoch: UInt64,
        sessionGeneration: UInt64,
        capabilityBits: UInt16,
        pairedAt: Date = .now
    ) throws {
        guard hostStaticPublicKey.count == 32,
              identityGeneration > 0,
              !deviceStaticKeyId.isEmpty,
              deviceStaticKeyId.utf8.count <= 128,
              !displayName.isEmpty,
              displayName.unicodeScalars.count <= 256 else {
            throw ControllerModelError.invalidHost
        }
        self.schemaVersion = Self.currentSchemaVersion
        self.id = id
        self.displayName = displayName
        self.route = route
        self.hostStaticPublicKey = hostStaticPublicKey
        self.deviceStaticKeyId = deviceStaticKeyId
        self.deviceId = deviceId
        self.identityGeneration = identityGeneration
        self.revocationEpoch = revocationEpoch
        self.sessionGeneration = sessionGeneration
        self.capabilityBits = capabilityBits
        self.pairedAt = pairedAt
    }

    var fingerprint: String {
        hostStaticPublicKey.map { String(format: "%02x", $0) }.joined()
    }
}

struct HostSummary: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let title: String
    let route: HostRoute
    let fingerprint: String
    let capabilityBits: UInt16
}

enum ControllerSessionOrigin: String, Codable, Hashable, Sendable {
    case terminal
    case managedAgent = "managed_agent"
    case observedAgent = "observed_agent"
    case unknown
}

enum ControllerSessionCapability: String, Codable, Hashable, Sendable {
    case observeSessions = "observe_sessions"
    case attachOutput = "attach_output"
    case sendInput = "send_input"
    case resize
    case respondToApproval = "respond_to_approval"
}

struct ControllerSessionSummary: Codable, Identifiable, Hashable, Sendable {
    let id: UUID
    let hostInstanceID: UUID?
    let origin: ControllerSessionOrigin
    let runtime: String?
    let capabilities: [ControllerSessionCapability]
    let title: String
    let project: String?
    let group: String?
    let lifecycle: String
    let activity: String?
    let occupantGeneration: UInt64?
    let lastOutputSequence: UInt64
    let hasWriter: Bool
    let unreadCount: UInt32

    init(
        id: UUID,
        hostInstanceID: UUID? = nil,
        origin: ControllerSessionOrigin = .unknown,
        runtime: String? = nil,
        capabilities: [ControllerSessionCapability] = [],
        title: String,
        project: String?,
        group: String?,
        lifecycle: String,
        activity: String?,
        occupantGeneration: UInt64?,
        lastOutputSequence: UInt64,
        hasWriter: Bool,
        unreadCount: UInt32
    ) {
        self.id = id
        self.hostInstanceID = hostInstanceID
        self.origin = origin
        self.runtime = runtime
        self.capabilities = capabilities
        self.title = title
        self.project = project
        self.group = group
        self.lifecycle = lifecycle
        self.activity = activity
        self.occupantGeneration = occupantGeneration
        self.lastOutputSequence = lastOutputSequence
        self.hasWriter = hasWriter
        self.unreadCount = unreadCount
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case hostInstanceID
        case origin
        case runtime
        case capabilities
        case title
        case project
        case group
        case lifecycle
        case activity
        case occupantGeneration
        case lastOutputSequence
        case hasWriter
        case unreadCount
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            id: try values.decode(UUID.self, forKey: .id),
            hostInstanceID: try values.decodeIfPresent(UUID.self, forKey: .hostInstanceID),
            origin: try values.decodeIfPresent(ControllerSessionOrigin.self, forKey: .origin) ?? .unknown,
            runtime: try values.decodeIfPresent(String.self, forKey: .runtime),
            capabilities: try values.decodeIfPresent([ControllerSessionCapability].self, forKey: .capabilities) ?? [],
            title: try values.decode(String.self, forKey: .title),
            project: try values.decodeIfPresent(String.self, forKey: .project),
            group: try values.decodeIfPresent(String.self, forKey: .group),
            lifecycle: try values.decode(String.self, forKey: .lifecycle),
            activity: try values.decodeIfPresent(String.self, forKey: .activity),
            occupantGeneration: try values.decodeIfPresent(UInt64.self, forKey: .occupantGeneration),
            lastOutputSequence: try values.decode(UInt64.self, forKey: .lastOutputSequence),
            hasWriter: try values.decode(Bool.self, forKey: .hasWriter),
            unreadCount: try values.decode(UInt32.self, forKey: .unreadCount)
        )
    }

    func validate() throws {
        guard !title.isEmpty,
              title.unicodeScalars.count <= ControllerCacheLimits.maxTitleScalars,
              lifecycle.utf8.count <= 64,
              runtime?.utf8.count ?? 0 <= 128,
              capabilities.count <= 5,
              Set(capabilities).count == capabilities.count,
              activity?.utf8.count ?? 0 <= 64,
              project?.unicodeScalars.count ?? 0 <= ControllerCacheLimits.maxTitleScalars,
              group?.unicodeScalars.count ?? 0 <= ControllerCacheLimits.maxTitleScalars else {
            throw ControllerModelError.invalidSession
        }
    }
}

struct SessionSummaryPage: Codable, Equatable, Sendable {
    let revision: UInt64
    let updateSequence: UInt64
    let sessions: [ControllerSessionSummary]
    let nextCursor: String?

    func validate(encodedBy encoder: JSONEncoder = JSONEncoder()) throws {
        guard revision > 0,
              updateSequence > 0,
              sessions.count <= ControllerCacheLimits.maxPageRecords,
              nextCursor?.utf8.count ?? 0 <= 256 else {
            throw ControllerModelError.resourceLimit
        }
        try sessions.forEach { try $0.validate() }
        guard try encoder.encode(self).count <= ControllerCacheLimits.maxPageBytes else {
            throw ControllerModelError.resourceLimit
        }
    }
}

struct ControllerFleetSnapshot: Equatable, Sendable {
    let revision: UInt64
    let updateSequence: UInt64
    let sessions: [ControllerSessionSummary]
}

enum ControllerConnectionState: Equatable, Sendable {
    case unpaired
    case pairing
    case sasReady(String)
    case pairedOffline
    case connecting
    case authenticating
    case syncing
    case readyReadOnly
    case revoked
    case incompatible
    case failed(ControllerFailure)
}

enum ControllerFailure: String, Codable, Error, Sendable {
    case cancelled
    case invalidOffer
    case offerExpired
    case sasMismatch
    case networkUnavailable
    case timedOut
    case authenticationFailed
    case keychainUnavailable
    case malformedResponse
    case sequenceGap
    case resourceLimit
    case storageUnavailable
    case pairingUncertain
}

struct ControllerViewState: Equatable, Sendable {
    let hosts: [HostSummary]
    let selectedHostID: String?
    let sessions: [ControllerSessionSummary]
    let connection: ControllerConnectionState
    let cacheUpdatedAt: Date?
    let isCachedReadOnly: Bool

    static let empty = Self(
        hosts: [],
        selectedHostID: nil,
        sessions: [],
        connection: .unpaired,
        cacheUpdatedAt: nil,
        isCachedReadOnly: false
    )
}

enum ControllerModelError: Error, Equatable {
    case invalidRoute
    case invalidHost
    case invalidSession
    case resourceLimit
}

enum ControllerCacheLimits {
    static let maxHosts = 16
    static let maxSessionsPerHost = 5_000
    static let maxSessionsGlobal = 10_000
    static let maxEncodedBytes = 8 * 1_024 * 1_024
    static let maxPageRecords = 1_000
    static let maxPageBytes = 1 * 1_024 * 1_024
    static let maxTitleScalars = 256
}

File: app/src/main/java/com/termirust/mobile/controller/ControllerModels.kt
package com.termirust.mobile.controller

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

object ControllerLimits {
    const val MAX_HOSTS = 16
    const val MAX_SESSIONS_PER_HOST = 5_000
    const val MAX_SESSIONS_TOTAL = 10_000
    const val MAX_CACHE_BYTES = 8 * 1_024 * 1_024
    const val MAX_PAGE_RECORDS = 1_000
    const val MAX_PAGE_BYTES = 1 * 1_024 * 1_024
    const val MAX_TITLE_CODE_POINTS = 256
}

@Serializable
data class HostRoute(
    val address: String,
    val port: Int,
) {
    init {
        require(address.isNotBlank() && address.toByteArray().size <= 255)
        require(port in 1..65_535)
    }
}

@Serializable
data class PairedHostRecord(
    @SerialName("schema_version") val schemaVersion: Int = 1,
    val id: String,
    @SerialName("display_name") val displayName: String,
    val route: HostRoute,
    @SerialName("host_static_public_key") val hostStaticPublicKey: String,
    @SerialName("device_static_key_id") val deviceStaticKeyId: String,
    @SerialName("device_id") val deviceId: String,
    @SerialName("identity_generation") val identityGeneration: Long,
    @SerialName("revocation_epoch") val revocationEpoch: Long,
    @SerialName("session_generation") val sessionGeneration: Long,
    @SerialName("capability_bits") val capabilityBits: Int,
    @SerialName("paired_at_millis") val pairedAtMillis: Long,
) {
    fun validate() {
        require(schemaVersion == 1)
        require(id.isNotBlank() && id.toByteArray().size <= 256)
        require(displayName.codePointCount() in 1..ControllerLimits.MAX_TITLE_CODE_POINTS)
        require(deviceStaticKeyId.toByteArray().size in 1..128)
        require(identityGeneration > 0 && revocationEpoch >= 0 && sessionGeneration > 0)
        require(capabilityBits in 0..0x1f)
    }
}

@Serializable
enum class ControllerSessionOrigin {
    @SerialName("terminal") TERMINAL,
    @SerialName("managed_agent") MANAGED_AGENT,
    @SerialName("observed_agent") OBSERVED_AGENT,
    @SerialName("unknown") UNKNOWN,
}

@Serializable
enum class ControllerSessionCapability {
    @SerialName("observe_sessions") OBSERVE_SESSIONS,
    @SerialName("attach_output") ATTACH_OUTPUT,
    @SerialName("send_input") SEND_INPUT,
    @SerialName("resize") RESIZE,
    @SerialName("respond_to_approval") RESPOND_TO_APPROVAL,
}

@Serializable
data class ControllerSessionSummary(
    val id: String,
    @SerialName("host_instance_id") val hostInstanceId: String? = null,
    val origin: ControllerSessionOrigin = ControllerSessionOrigin.UNKNOWN,
    val runtime: String? = null,
    val capabilities: List<ControllerSessionCapability> = emptyList(),
    val title: String,
    val project: String? = null,
    val group: String? = null,
    val lifecycle: String,
    val activity: String? = null,
    @SerialName("occupant_generation") val occupantGeneration: Long? = null,
    @SerialName("last_output_sequence") val lastOutputSequence: Long,
    @SerialName("has_writer") val hasWriter: Boolean,
    @SerialName("unread_count") val unreadCount: Int,
) {
    fun validate() {
        require(id.isUuid())
        require(hostInstanceId == null || hostInstanceId.isUuid())
        require(runtime == null || runtime.toByteArray().size in 1..128)
        require(capabilities.size <= 5 && capabilities.toSet().size == capabilities.size)
        require(title.codePointCount() in 1..ControllerLimits.MAX_TITLE_CODE_POINTS)
        require(project == null || project.codePointCount() <= ControllerLimits.MAX_TITLE_CODE_POINTS)
        require(group == null || group.codePointCount() <= ControllerLimits.MAX_TITLE_CODE_POINTS)
        require(lifecycle.toByteArray().size in 1..64)
        require(activity == null || activity.toByteArray().size <= 64)
        require(occupantGeneration == null || occupantGeneration > 0)
        require(lastOutputSequence >= 0 && unreadCount >= 0)
    }
}

@Serializable
data class ControllerFleetSnapshot(
    val revision: Long,
    @SerialName("update_sequence") val updateSequence: Long,
    val sessions: List<ControllerSessionSummary>,
) {
    fun validate() {
        require(revision > 0 && updateSequence > 0)
        require(sessions.size <= ControllerLimits.MAX_SESSIONS_PER_HOST)
        sessions.forEach(ControllerSessionSummary::validate)
        require(sessions.map { it.id }.toSet().size == sessions.size)
    }
}

@Serializable
data class CachedHostFleet(
    val host: PairedHostRecord,
    val snapshot: ControllerFleetSnapshot,
    @SerialName("updated_at_millis") val updatedAtMillis: Long,
    @SerialName("last_viewed_at_millis") val lastViewedAtMillis: Long,
)

@Serializable
data class ControllerCacheDocument(
    @SerialName("schema_version") val schemaVersion: Int = 1,
    val hosts: Map<String, CachedHostFleet> = emptyMap(),
)

sealed interface ControllerConnectionState {
    data object Unpaired : ControllerConnectionState
    data object Pairing : ControllerConnectionState
    data class SasReady(val sas: String, val fingerprintSuffix: String) : ControllerConnectionState
    data object PairedOffline : ControllerConnectionState
    data object Connecting : ControllerConnectionState
    data object Authenticating : ControllerConnectionState
    data object Syncing : ControllerConnectionState
    data object ReadyReadOnly : ControllerConnectionState
    data object Revoked : ControllerConnectionState
    data object Incompatible : ControllerConnectionState
    data class Failed(val code: String) : ControllerConnectionState
}

data class ControllerUiState(
    val hosts: List<PairedHostRecord> = emptyList(),
    val selectedHostId: String? = null,
    val sessions: List<ControllerSessionSummary> = emptyList(),
    val connection: ControllerConnectionState = ControllerConnectionState.Unpaired,
    val cachedAtMillis: Long? = null,
    val cachedReadOnly: Boolean = false,
    val activeTerminal: ControllerTerminalUiState? = null,
)

data class ControllerTerminalUiState(
    val hostTitle: String,
    val sessionTitle: String,
    val attachState: ReadOnlyAttachState,
    val screen: BoundedTerminalSnapshot,
    val outputSequence: Long,
    val hasWriterElsewhere: Boolean = false,
    val writerLease: WriterLeaseState = WriterLeaseState.None,
    val writerMessage: String? = null,
    val pendingPasteBytes: Int = 0,
    val supportsWriter: Boolean = false,
    val supportsResize: Boolean = false,
    val privacyCovered: Boolean = false,
)

data class ControllerPairingChallenge(
    val sas: String,
    val fingerprintSuffix: String,
    val route: HostRoute,
    val expiresAtMillis: Long,
)

private fun String.codePointCount(): Int = codePointCount(0, length)

private fun String.isUuid(): Boolean = runCatching { java.util.UUID.fromString(this) }.isSuccess
